mu4/extensions/internal/extensionscontroller.cpp

```diff
@@ -0,0 +1,372 @@
+        extension.fileName = value.value("file_name").toString();
+        extension.fileSize = value.value("file_size").toDouble();
+        extension.version = QVersionNumber::fromString(value.value("version").toString());
+        extension.types = {};
+
+        result.val.insert(key, extension);
+    }
+
+    return result;
+}
+
+bool ExtensionsController::isExtensionExists(const QString& extensionCode) const
+{
+    return fsOperation()->exists(configuration()->extensionsSharePath() + "/" + extensionCode);
+}
+
+RetVal<ExtensionsHash> ExtensionsController::correctExtensionsStates(ExtensionsHash& extensions) const
+{
+    RetVal<ExtensionsHash> result;
+    bool isNeedUpdate = false;
+
+    for (Extension& extension: extensions) {
+        if (extension.status == ExtensionStatus::Installed && !isExtensionExists(extension.code)) {
+            extension.status = ExtensionStatus::NoInstalled;
+            isNeedUpdate = true;
+        }
+
+        extension.types = extensionTypes(extension.code);
+    }
+
+    if (isNeedUpdate) {
+        Ret ret = configuration()->setExtensions(extensions);
+        if (!ret) {
+            result.ret = ret;
+            return result;
+        }
+    }
+
+    result.ret = make_ret(Err::NoError);
+    result.val = extensions;
+    return result;
+}
+
+RetVal<QString> ExtensionsController::downloadExtension(const QString& extensionCode,
+                                                        async::Channel<ExtensionProgress>& progressChannel) const
+{
+    RetVal<QString> result;
+
+    ValCh<ExtensionsHash> extensions = configuration()->extensions();
+    QString fileName = extensions.val.value(extensionCode).fileName;
+
+    QString extensionArchivePath = configuration()->extensionsDataPath() + "/" + fileName;
+
+    QBuffer buff;
+    INetworkManagerPtr networkManagerPtr = networkManagerCreator()->makeNetworkManager();
+
+    async::Channel<Progress> downloadChannel = networkManagerPtr->downloadProgressChannel();
+    downloadChannel.onReceive(new deto::async::Asyncable(), [&progressChannel](const Progress& progress) {
+        progressChannel.send(ExtensionProgress(qtrc("extensions", "Downloading..."), progress.current,
+                                                     progress.total));
+    });
+
+    Ret getExtension = networkManagerPtr->get(configuration()->extensionsFileServerUrl().toString() + fileName, &buff);
+
+    if (!getExtension) {
+        LOGE() << "Error save file";
+        result.ret = make_ret(Err::ErrorLoadingExtension);
+        return result;
+    }
+
+    QFile file(extensionArchivePath);
+    file.open(QIODevice::WriteOnly);
+    file.write(buff.data());
+    file.close();
+
+    result.ret = make_ret(Err::NoError);
+    result.val = extensionArchivePath;
+    return result;
+}
+
+Ret ExtensionsController::removeExtension(const QString& extensionCode) const
+{
+    QString extensionPath = configuration()->extensionsSharePath() + "/" + extensionCode;
```

Comment:
I see it at least the second time
Better to do this:
```
QString extensionPath = configuration()->makeExtensionPath(extensionCode);
```
And all these things are better done this way.
This eliminates duplication, we can easily see which paths are where.